doc/man/zorp.8
'\" t
.\"     Title: zorp 
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 10/08/2014
.\"    Manual: [FIXME: manual]
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "ZORP" "8" "10/08/2014" "[FIXME: source]" "[FIXME: manual]"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
zorp_ \- Zorp Firewall Suite
.SH "SYNOPSIS"
.HP \w'\fBzorp\fR\ 'u
\fBzorp\fR [options]
.SH "DESCRIPTION"
.PP
The
\fBzorp\fR
command is the main entry point for a Zorp instance, and as such it is generally called by
zorpctl(8)
with command line parameters specified in
instances\&.conf(5)\&.
.SH "OPTIONS"
.PP
\fB\-\-version\fR or \fB\-V\fR
.RS 4
Display version number and compilation information\&.
.RE
.PP
\fB\-\-as <name>\fR or \fB\-a <name>\fR
.RS 4
Set instance name to <name>\&. Instance names may consist of the characters [a\-zA\-Z0\-9_] and must begin with a letter\&. Log messages of this instance are prefixed with this name\&.
.RE
.PP
\fB\-\-also\-as <name>\fR or \fB\-A <name>\fR
.RS 4
Add a secondary instance named <name>\&. Secondary instances share the same Zorp process but they have a separate section in the configuration file\&.
.RE
.PP
\fB\-\-policy <name>\fR or \fB\-p <name>\fR
.RS 4
Use the file called <name> as policy\&. This file must be a valid policy file\&.
.RE
.PP
\fB\-\-verbose <verbosity>\fR or \fB\-v <verbosity>\fR
.RS 4
Set verbosity level to <verbosity>, or if <verbosity> is omitted increment it by one\&. Default the verbosity level is 3; possible values are 0\-10\&.
.RE
.PP
\fB\-\-pidfile <pidfile>\fR or \fB\-P <pidfile>\fR
.RS 4
Set path to the PID file where the pid of the main process is stored\&.
.RE
.PP
\fB\-\-foreground\fR or \fB\-F\fR
.RS 4
Do not daemonize, run in the foreground\&.
.RE
.PP
\fB\-\-process\-mode <mode>\fR
.RS 4
Set processing mode to one of background, safe\-background or foreground\&.
.RE
.PP
\fB\-\-no\-syslog\fR or \fB\-l\fR
.RS 4
Send log messages to the standard output instead of syslog\&.
.RE
.PP
\fB\-\-log\-tags\fR or \fB\-T\fR
.RS 4
Prepend log category and log level to each message\&.
.RE
.PP
\fB\-\-log\-escape\fR
.RS 4
Escape non\-printable characters to avoid binary log files\&. Each character less than 0x20 and greater than 0x7F are escaped in the form <XX>\&.
.RE
.PP
\fB\-\-log\-spec <spec>\fR or \fB\-s <spec>\fR
.RS 4
Set verbosity mask on a per category basis\&. Each log message has an assigned multi\-level category, where levels are separated by a dot\&. For example, HTTP requests are logged under
\fIhttp\&.request\fR\&.
\fB<spec>\fR
is a comma separated list of log specifications\&. A single log specification consists of a wildcard matching log category, a colon, and a number specifying the verbosity level of that given category\&. Categories match from left to right\&. E\&.g\&.:
\fB\-\-logspec \*(Aqhttp\&.*:5,core:3\*(Aq\fR\&. The last matching entry will be used as the verbosity of the given category\&. If no match is found the default verbosity specified with
\fB\-\-verbose\fR
is used\&.
.RE
.PP
\fB\-\-threads <num>\fR or \fB\-t <num>\fR
.RS 4
Set the maximum number of threads that can be used in parallel by this Zorp instance\&.
.RE
.PP
\fB\-\-idle\-threads <num>\fR or \fB\-I\fR
.RS 4
Set the maximum number of idle threads; this option has effect only if threadpools are enabled (see the option
\fB\-\-threadpools\fR)\&.
.RE
.PP
\fB\-\-threadpools\fR or \fB\-O\fR
.RS 4
Enable the use of threadpools, which means that threads associated with sessions are not automatically freed, only if the maximum number of idle threads is exceeded\&.
.RE
.PP
\fB\-\-user <user>\fR or \fB\-u <user>\fR
.RS 4
Switch to the supplied user after starting up\&.
.RE
.PP
\fB\-\-group <group>\fR or \fB\-g <group>\fR
.RS 4
Switch to the supplied group after starting up\&.
.RE
.PP
\fB\-\-chroot <dir>\fR or \fB\-R <dir>\fR
.RS 4
Change root to the specified directory before reading the configuration file\&. The directory must be set up accordingly\&.
.RE
.PP
\fB\-\-caps <caps>\fR or \fB\-C <caps>\fR
.RS 4
Switch to the supplied set of capabilities after starting up\&. This should contain the required capabilities in the permitted set\&. For the syntax of capability description see the man page
cap_from_text(3)\&.
.RE
.PP
\fB\-\-no\-caps\fR or \fB\-N\fR
.RS 4
Do not change capabilities at all\&.
.RE
.PP
\fB\-\-crypto\-engine <engine>\fR or \fB\-E <engine>\fR
.RS 4
Set the OpenSSL crypto engine to be used for hardware accelerated crypto support\&.
.RE
.PP
\fB\-\-stack\-size <size>\fR or \fB\-S <size>\fR
.RS 4
Set the maximum stack size used by threads\&. Note that the maximum number of parallel threads is influenced by the size specified here\&. The default stack size is 512 KB, the maximum you can set is 8192 KB\&.
.RE
.SH "FILES"
.PP
/etc/zorp/
.PP
/etc/zorp/policy\&.py
.PP
/etc/zorp/instances\&.conf
<xi:include></xi:include><xi:include></xi:include>

debian/changelog
zorp (5.0.0) zorp50gpl; urgency=low

  * New upstream version.

 -- BalaBit Development Team <devel@balabit.hu>  Wed, 08 Oct 2014 15:48:49 +0200

zorp (3.9.0-0.1) unstable; urgency=low

  * removed unnecessary shlib files
  * fixed file name format of .files
  * added .files file for python-kzorp package
  * added new proxy related python files to the zorp package
  * fixed package name formats
  * removed explicit dependecy against zorp-modules
  * fixed build dependencies of package zorp

 -- Szilárd Pfeiffer <szilard.pfeiffer@balabit.com>  Thu, 03 Feb 2010 15:28:28 +0100

zorp (3.3.6-1.1) unstable; urgency=low

  * Non-maintainer upload
  * Add to libzorp0: "Replaces: "libzorp2; Breaks: libzorp2" to fix error
    when upgrading from lenny (closes: #601991). Thanks to Marcos Marado
    for the hint.

 -- Ralf Treinen <treinen@debian.org>  Thu, 11 Nov 2010 15:28:28 +0100

zorp (3.3.6-1) unstable; urgency=low

  * New upstream version.

 -- SZALAY Attila <sasa@debian.org>  Tue, 29 Jun 2010 17:18:42 +0200

zorp (3.0.8-0.5) unstable; urgency=high

  * Non-maintainer upload.
  * Force using python2.4. Closes: #506700.

 -- Thomas Viehmann <tv@beamnet.de>  Sun, 23 Nov 2008 21:54:36 +0100

zorp (3.0.8-0.4) unstable; urgency=low

  * Non-maintainer upload.
  * Use PyObject_Del instead of PyMem_DEL. (Closes: #469012).
  * zorp suggests zorp-doc. (Closes: #392141).
  * LSB formatted init script. (Closes: #469115).
    + Thanks to Petter Reinholdtsen.

 -- Barry deFreese <bddebian@comcast.net>  Mon, 03 Mar 2008 10:44:13 -0500

zorp (3.0.8-0.3) unstable; urgency=high

  * Non-maintainer upload.
  * High-urgency upload for RC bugfix.
  * debian/rules: Explicitly run aclocal-1.7 instead of aclocal (FTBFS with
    aclocal-1.9).
  * configure.in: Set datadir and libdir to ${prefix}/share and ${prefix}/lib.
    This is fixed upstream in version 3.0.14b with more intrusive changes in
    the build system. Note that ./configure --datadir and --libdir won't work.
    Closes: #413933
  * Thanks to Nicolas Francois for the patch.
  * Also fix the clean target to use make distclean, not just make
    clean.
  * The previous change requires setting ZORP_PACKAGE_LICENSE=gpl when
    calling configure to avoid another FTBFS.

 -- Steve Langasek <vorlon@debian.org>  Sun, 18 Mar 2007 20:54:44 -0700

zorp (3.0.8-0.2) unstable; urgency=high

  * Non-maintainer upload to fix release critical bug.
  * Downgrade Depends of zorp-modules on zorp to a Recommends;
    remove cyclic dependency that broke purge (Closes: #389769).

 -- Thijs Kinkhorst <thijs@debian.org>  Tue, 10 Oct 2006 15:58:17 +0200

zorp (3.0.8-0.1) unstable; urgency=low

  * Non-maintainer upload.
  * Update package to the last python policy (Closes: 380994).
  * Make the package binNMUable (use ${source:Version}).

 -- Pierre Habouzit <madcoder@debian.org>  Sun,  3 Sep 2006 23:40:28 +0200

zorp (3.0.8) unstable; urgency=low

  * New upstream version.

 -- SZALAY Attila <sasa@debian.org>  Tue, 24 Jan 2006 19:35:29 +0000

zorp (3.0.5) unstable; urgency=low

  * New upstream version.

 -- SZALAY Attila <sasa@debian.org>  Mon, 31 Oct 2005 16:07:32 +0000

zorp (2.0.9-10) unstable; urgency=high

  * Fixed FTBFS bug by Christian Hammers <ch@debian.org>. Closes: #304666

 -- SZALAY Attila <sasa@debian.org>  Wed, 25 May 2005 19:42:51 +0200

zorp (2.0.9-9) unstable; urgency=low

  * Fixed invalid lvalue in test_registry. Closes: #290446

 -- Magosányi Árpád (mag) <mag@debian.org>  Tue, 18 Jan 2005 10:17:54 +0000

zorp (2.0.9-8) unstable; urgency=low

  * tests are disabled for now. Closes: #280947

 -- Magosányi Árpád (mag) <mag@debian.org>  Mon, 22 Nov 2004 21:53:00 +0100

zorp (2.0.9-6) unstable; urgency=low

  * makes tests runnable. Closes: #275978

 -- Magosányi Árpád (mag) <mag@debian.org>  Sat, 30 Oct 2004 15:12:05 +0200

zorp (2.0.9-5) unstable; urgency=low

  * fixes python2.3 related crash. Closes: 262233
  * we now have tests

 -- Magosányi Árpád (mag) <mag@debian.org>  Mon, 13 Sep 2004 16:47:24 +0200

zorp (2.0.9-4) unstable; urgency=low

  * fixed libzorpll build-dependency

 -- Magosányi Árpád (mag) <mag@debian.org>  Sun, 25 Jul 2004 06:22:01 +0000

zorp (2.0.9-3) unstable; urgency=low

  * made libzorpll dependencies more explicit. Closes: #255734
  * fixed instances.conf manpage section
  * dropped g++ dependency, which has been brought in because an autoconf
    problem
  * added zorp-config manpage to libzorp2

 -- Magosányi Árpád (mag) <mag@debian.org>  Fri,  9 Jul 2004 06:27:28 +0000

zorp (2.0.9-2) unstable; urgency=low

  * Build with newer libzorpll
  * Changelog and control are properly UTF-8

 -- Magosányi Árpád (mag) <mag@debian.org>  Sun, 13 Jun 2004 23:17:34 +0200

zorp (2.0.9-1) unstable; urgency=low

  * Merged in upstream changes

 -- Magosányi Árpád (mag) <mag@debian.org>  Sat, 12 Jun 2004 01:04:49 +0200

zorp (2.0.8-3) unstable; urgency=low

  * Signedness bug. Closes: #177287
  * Fixed the init script. Closes: #240355
  * Fixed the uploader field in debian/control

 -- Magosányi Árpád (mag) <mag@debian.org>  Fri, 16 Apr 2004 14:51:33 +0000

zorp (2.0.8-2) unstable; urgency=low

  * New upstream version
  * 2.0.2-2 closed some bugs. Closes: #220911
  * Build with the correct python version for python-extclass. (hopefully) Closes: #220916

 -- Magosányi Árpád (mag) <mag@debian.org>  Fri, 12 Mar 2004 14:46:50 +0100

zorp (2.0.8-1) unstable; urgency=low

  * Unreleased

 -- Magosányi Árpád (mag) <mag@debian.org>  Wed, 10 Dec 2003 14:58:52 +0100

zorp (2.0.7-2) unstable; urgency=low

  * install instances.conf Closes: #220911
  * sets ulimit to 65532. Works around #221257

 -- Magosányi Árpád (mag) <mag@debian.org>  Tue,  2 Dec 2003 12:18:12 +0100

zorp (2.0.7-1) unstable; urgency=low

  * new upstream version
  * regenerate all generated files, brings in build dependency on autoconf
  * rearranged some includes to make gcc happy

 -- Magosányi Árpád (mag) <mag@debian.org>  Wed,  5 Nov 2003 01:35:44 +0100


pylib/Zorp/Zorp.py
############################################################################
##
## Copyright (c) 2000-2014 BalaBit IT Ltd, Budapest, Hungary
##
## This program is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License
## as published by the Free Software Foundation; either version 2
## of the License, or (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program; if not, write to the Free Software
## Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
##
############################################################################

"""
<module maturity="stable">
  <summary>
    Module defining interface to the Zorp core entry points.
  </summary>
  <description>
    <para>
      This module defines global constants (e.g., <parameter>TRUE</parameter> and
      <parameter>FALSE</parameter>) used by other Zorp components, and interface
      entry points to the Zorp core.
    </para>
  </description>
  <metainfo>
    <enums>
      <enum maturity="stable" id="enum.zorp.z">
        <description>
          Values returned by event handlers.
        </description>
        <item><name>ZV_UNSPEC</name></item>
        <item><name>ZV_ACCEPT</name></item>
        <item><name>ZV_DENY</name></item>
        <item><name>ZV_REJECT</name></item>
        <item><name>ZV_ABORT</name></item>
        <item><name>ZV_DROP</name></item>
        <item><name>ZV_POLICY</name></item>
      </enum>
     <enum maturity="stable" id="zorp.proto.id">
        <description>
          The network protocol used in the server-side connection.
        </description>
        <item>
          <name>ZD_PROTO_AUTO</name>
          <description>
            Use the protocol that is used on the client side.
          </description>
        </item>
        <item>
          <name>ZD_PROTO_TCP</name>
          <description>
            Use the TCP protocol on the server side.
          </description>
        </item>
        <item>
          <name>ZD_PROTO_UDP</name>
          <description>
            Use the UDP protocol on the server side.
          </description>
        </item>
      </enum>
      <enum maturity="stable" id="enum.zorp.forge_port">
      <description>
        Options defining the source port of the server-side connection.
      </description>
      <item>
        <name>Z_PORT_ANY</name>
        <description>
          Selected a random port between 1024
              and 65535. This is the default behavior of every router.
        </description>
      </item>
      <item>
        <name>Z_PORT_GROUP</name>
        <description>
          Select a random port in the same group as the port used by
          the client. The following groups are defined:
          <parameter>0-513</parameter>, <parameter>514-1024</parameter>,
          <parameter>1025-</parameter>.
        </description>
      </item>
      <item>
        <name>Z_PORT_EXACT</name>
        <description>
          Use the same port as the client.
        </description>
      </item>
      <item>
        <name>Z_PORT_RANDOM</name>
        <description>
          Select a random port using a cryptographically secure function.
        </description>
      </item>
      </enum>
      <enum maturity="stable" id="enum.zorp.bacl">
        <description>basic acl tags</description>
        <item><name>Z_BACL_REQUIRED</name></item>
        <item><name>Z_BACL_SUFFICIENT</name></item>
      </enum>
      <enum maturity="stable" id="enum.zorp.af">
        <description>address families</description>
        <item><name>AF_UNSPEC</name></item>
        <item><name>AF_INET</name></item>
        <item><name>AF_INET6</name></item>
      </enum>
      <enum maturity="stable" id="enum.zorp.stack">
        <description></description>
        <item><name>Z_STACK_PROXY</name></item>
        <item><name>Z_STACK_PROGRAM</name></item>
        <item><name>Z_STACK_REMOTE</name></item>
        <item><name>Z_STACK_PROVIDER</name></item>
        <item><name>Z_STACK_PROXY_IN_SESSION</name></item>
      </enum>
      <enum maturity="stable" id="enum.zorp.logical">
        <description>logical operators</description>
        <item><name>Z_NOT</name></item>
        <item><name>Z_AND</name></item>
        <item><name>Z_OR</name></item>
        <item><name>Z_XOR</name></item>
        <item><name>Z_EQ</name></item>
        <item><name>Z_NE</name></item>
      </enum>
    </enums>
    <actiontuples>
      <actiontuple maturity="stable" id="action.zorp.stack" action_enum="enum.zorp.stack">
        <description>
        Stacking options.
        </description>
        <tuple action="Z_STACK_PROXY">
          <args>
            <class filter="proxy"/>
          </args>
          <description>
          Stack a proxy.
          </description>
        </tuple>
        <tuple action="Z_STACK_PROGRAM">
          <args>
            <string/>
          </args>
          <description>
          Stack an external program.
          </description>
        </tuple>
        <tuple action="Z_STACK_REMOTE">
          <args>
            <tuple>
              <sockaddr/>
              <string/>
            </tuple>
          </args>
          <description>
          Stack a remote destination.
          </description>
        </tuple>
        <tuple action="Z_STACK_PROVIDER">
          <args>
            <tuple>
              <class filter="stackingprov" existing="yes"/>
              <string/>
            </tuple>
          </args>
          <description>
          Stack a Stacking Provider.
          </description>
        </tuple>
      </actiontuple>
    </actiontuples>
    <constants>
      <constantgroup maturity="stable" id="const.zorp.glob">
        <description>global variables</description>
        <item><name>firewall_name</name><value>"zorp"</value></item>
      </constantgroup>
      <constantgroup maturity="stable" id="const.zorp.core">
        <description>Core message tags</description>
        <item><name>CORE_SESSION</name><value>"core.session"</value></item>
        <item><name>CORE_DEBUG</name><value>"core.debug"</value></item>
        <item><name>CORE_ERROR</name><value>"core.error"</value></item>
        <item><name>CORE_POLICY</name><value>"core.policy"</value></item>
        <item><name>CORE_MESSAGE</name><value>"core.message"</value></item>
        <item><name>CORE_AUTH</name><value>"core.auth"</value></item>
      </constantgroup>
    </constants>
  </metainfo>
</module>
"""

firewall_name = "zorp" # obsolete, not used anymore

import traceback
import sys
import errno
import socket
import Config

config = Config

import Common

CORE_SESSION = Common.CORE_SESSION
CORE_DEBUG = Common.CORE_DEBUG
CORE_ERROR = Common.CORE_ERROR
CORE_POLICY = Common.CORE_POLICY
CORE_MESSAGE = Common.CORE_MESSAGE
CORE_AUTH = Common.CORE_AUTH
CORE_INFO = Common.CORE_INFO
CORE_ALERTING = Common.CORE_ALERTING

# return values returned by event handlers
ZV_UNSPEC         = 0
ZV_ACCEPT         = 1
ZV_DENY           = 2
ZV_REJECT         = 3
ZV_ABORT          = 4
ZV_DROP           = 5
ZV_POLICY         = 6
ZV_ERROR          = 7

# Legacy names
Z_UNSPEC         = ZV_UNSPEC
Z_ACCEPT         = ZV_ACCEPT
Z_DENY           = ZV_DENY
Z_REJECT         = ZV_REJECT
Z_ABORT          = ZV_ABORT
Z_DROP           = ZV_DROP
Z_POLICY         = ZV_POLICY
Z_ERROR          = ZV_ERROR

# dispatched protocols
ZD_PROTO_AUTO = 0
ZD_PROTO_TCP  = 1
ZD_PROTO_UDP  = 2

ZD_PROTO_NAME = (
   "AUTO",    # ZD_PROTO_AUTO
   "TCP",     # ZD_PROTO_TCP
   "UDP",     # ZD_PROTO_UDP
)

# port allocation values
Z_PORT_ANY = -1
Z_PORT_GROUP = -2
Z_PORT_EXACT = -3
Z_PORT_RANDOM = -4

# basic acl tags
Z_BACL_REQUIRED = 1
Z_BACL_SUFFICIENT = 2

# stack types
Z_STACK_PROXY = 1
Z_STACK_PROGRAM = 2
Z_STACK_REMOTE = 3
Z_STACK_PROVIDER = 4
Z_STACK_CUSTOM = 5
Z_STACK_PROXY_IN_SESSION = 6

# proxy priorities
Z_PROXY_PRI_LOW = 0
Z_PROXY_PRI_NORMAL = 1
Z_PROXY_PRI_HIGH = 2
Z_PROXY_PRI_URGENT = 3

# boolean values
FALSE = 0
TRUE = 1

# address families
AF_UNSPEC = socket.AF_UNSPEC
AF_INET = socket.AF_INET
AF_INET6 = socket.AF_INET6

# logical operators
Z_NOT  = "Z_NOT"
Z_AND  = "Z_AND"
Z_OR   = "Z_OR"
Z_XOR  = "Z_XOR"
Z_EQ   = "Z_EQ"
Z_NE   = "Z_XOR"

Z_SZIG_TYPE_LONG = 1
Z_SZIG_TYPE_TIME = 2
Z_SZIG_TYPE_STRING = 3
Z_SZIG_TYPE_PROPS = 4
Z_SZIG_TYPE_CONNECTION_PROPS = 5

Z_SZIG_THREAD_START = 0
Z_SZIG_THREAD_STOP = 1
Z_SZIG_TICK = 2
Z_SZIG_COUNTED_IP = 3
Z_SZIG_CONNECTION_PROPS = 4
Z_SZIG_CONNECTION_STOP = 5
Z_SZIG_AUDIT_START = 6
Z_SZIG_AUDIT_STOP = 7
Z_SZIG_RELOAD = 8
Z_SZIG_AUTH_PENDING_BEGIN = 9
Z_SZIG_AUTH_PENDING_FINISH = 10
Z_SZIG_SERVICE_COUNT = 11
Z_SZIG_CONNECTION_START = 12

Z_KEEPALIVE_NONE   = 0
Z_KEEPALIVE_CLIENT = 1
Z_KEEPALIVE_SERVER = 2
Z_KEEPALIVE_BOTH   = 3

Z_SSL_VERIFY_NONE               = 0
Z_SSL_VERIFY_OPTIONAL_UNTRUSTED = 1
Z_SSL_VERIFY_OPTIONAL_TRUSTED   = 2
Z_SSL_VERIFY_REQUIRED_UNTRUSTED = 3
Z_SSL_VERIFY_REQUIRED_TRUSTED   = 4

import Globals

def init(names, virtual_name, is_master):
    """
    <function internal="yes">
      <summary>
        Default init() function provided by Zorp
      </summary>
      <description>
        This function is a default <function>init()</function> calling the init function
        identified by the <parameter>name</parameter> argument. This way several Zorp
        instances can use the same policy file.
      </description>
      <metainfo>
        <attributes>
          <attribute maturity="stable">
            <name>names</name>
            <type></type>
            <description>Names (instance name and also-as names) of this instance.</description>
          </attribute>
          <attribute maturity="stable">
            <name>virtual_name</name>
            <type>string</type>
            <description>
              Virtual instance name of this process. If a Zorp instance is backed by multiple
              Zorp processes using the same configuration each process has a unique virtual
              instance name that is used for SZIG communication, PID file creation, etc.
            </description>
          </attribute>
          <attribute>
            <name>is_master</name>
            <type>int</type>
            <description>
              TRUE if Zorp is running in master mode, FALSE for slave processes. Each Zorp instance
              should have exactly one master process and an arbitrary number of slaves.
            </description>
          </attribute>
        </attributes>
      </metainfo>
    </function>
    """
    import __main__
    import SockAddr, KZorp, Matcher, Rule
    import kzorp.netlink
    import kzorp.communication
    import errno

    # miscelanneous initialization
    if config.audit.encrypt_certificate_file:
        try:
            config.audit.encrypt_certificate = open(config.audit.encrypt_certificate_file, 'r').read()
        except IOError:
            log(None, CORE_ERROR, 1, "Error reading audit encryption certificate; file='%s'", (config.audit.encrypt_certificate_file))

    if config.audit.encrypt_certificate_list_file:
        try:
            config.audit.encrypt_certificate_list = [ ]
            for list in config.audit.encrypt_certificate_list_file:
                newlist = [ ]
                for file in list:
                    try:
                        newlist.append( open(file, 'r').read() )
                    except IOError:
                        log(None, CORE_ERROR, 1, "Error reading audit encryption certificate; file='%s'", (file))
                config.audit.encrypt_certificate_list.append( newlist )
        except TypeError:
            log(None, CORE_ERROR, 1, "Error iterating encryption certificate file list;")

    if config.audit.encrypt_certificate_list == None and config.audit.encrypt_certificate:
        config.audit.encrypt_certificate_list = [ [ config.audit.encrypt_certificate ] ]

    if config.audit.sign_private_key_file:
        try:
            config.audit.sign_private_key = open(config.audit.sign_private_key_file, 'r').read()
        except IOError:
            log(None, CORE_ERROR, 1, "Error reading audit signature's private key; file='%s'", (config.audit.sign_private_key_file))

    if config.audit.sign_certificate_file:
        try:
            config.audit.sign_certificate = open(config.audit.sign_certificate_file, 'r').read()
        except IOError:
            log(None, CORE_ERROR, 1, "Error reading audit signature's certificate; file='%s'", (config.audit.sign_certificate_file))

    Globals.rules = Rule.RuleSet()

    if config.options.kzorp_enabled:
        # ping kzorp to see if it's there
        try:
            h = kzorp.communication.Handle()
            Globals.kzorp_available = True
        except:
            Globals.kzorp_available = False
            log(None, CORE_ERROR, 0, "Error pinging KZorp, it is probably unavailable; exc_value='%s'" % (sys.exc_value))

    Globals.instance_name = names[0]
    for i in names:
        try:
            func = getattr(__main__, i)
        except AttributeError:
            ## LOG ##
            # This message indicates that the initialization function of
            # the given instance was not found in the policy file.
            ##
            log(None, CORE_ERROR, 0, "Instance definition not found in policy; instance='%s'", (names,))
            return FALSE
        func()

    Matcher.validateMatchers()

    if Globals.kzorp_available:
        try:
            KZorp.downloadKZorpConfig(names[0], is_master)
        except:
            ## LOG ##
            # This message indicates that downloading the necessary information to the
            # kernel-level KZorp subsystem has failed.
            ##
            log(None, CORE_ERROR, 0, "Error downloading KZorp configuration, Python traceback follows; error='%s'" % (sys.exc_value))
            for s in traceback.format_tb(sys.exc_traceback):
                for l in s.split("\n"):
                    if l:
                        log(None, CORE_ERROR, 0, "Traceback: %s" % (l))

            # if kzorp did respond to the ping, the configuration is erroneous -- we die here so the user finds out
            return FALSE

    return TRUE


def deinit(names, virtual_name):
    """
    <function internal="yes">
    </function>
    """
    ## LOG ##
    # This message reports that the given instance is stopping.
    ##
    log(None, CORE_DEBUG, 6, "Deinitialization requested for instance; name='%s'", (names[0],))
    for i in Globals.deinit_callbacks:
        i()

def purge():
    """
    <function internal="yes">
    </function>
    """
    pass

def cleanup(names, virtual_name, is_master):
    """
    <function internal="yes">
    </function>
    """
    import KZorp
    ## LOG ##
    # This message reports that the given instance is freeing its external
    # resources (for example its kernel-level policy objects).
    ##
    log(None, CORE_DEBUG, 6, "Cleaning up instance; name='%s'", (names,))

    if is_master and Globals.kzorp_available and config.options.kzorp_enabled:
        try:
            KZorp.flushKZorpConfig(names[0])
        except:
            ## LOG ##
            # This message indicates that flushing the instance-related information in the
            # kernel-level KZorp subsystem has failed.
            ##
            log(None, CORE_ERROR, 0, "Error flushing KZorp configuration; error='%s'" % (sys.exc_value))
            for s in traceback.format_tb(sys.exc_traceback):
                for l in s.split("\n"):
                    if l:
                        log(None, CORE_ERROR, 4, "Traceback: %s" % (l))

def notify(event, params):
    """<function internal="yes">
    </function>
    """
    if Globals.notification_policy:
        return Globals.notification_policy.notify(event, params)



## NOLOG ##

def log(sessionid, logclass, verbosity, msg, args=None):
    """
    <function maturity="stable">
      <summary>
        Function to send a message to the system log.
      </summary>
      <description>
        <para>
          This function can be used to send a message to the system log.
        </para>
      </description>
      <metainfo>
        <arguments>
          <argument>
           <name>sessionid</name>
           <type><string/></type>
           <description>The ID of the session the message belongs to.</description>
          </argument>
          <argument>
            <name>logclass</name>
            <type><string/></type>
            <description>Hierarchical log class as described in the <emphasis>zorp(8)</emphasis> manual page</description>
          </argument>
          <argument>
            <name>verbosity</name>
            <type><integer/></type>
            <description>Verbosity level of the message.</description>
          </argument>
          <argument>
            <name>msg</name>
            <type><string/></type>
            <description>The message text.</description>
          </argument>
          <argument>
            <name>args</name>
            <type><string/></type>
            <description>Optional printf-style argument tuple added to the message.</description>
          </argument>
        </arguments>
      </metainfo>
    </function>
    """
    Common.log(sessionid, logclass, verbosity, msg, args=None)
